authentication-service/src/main.rs

```diff
@@ -130,81 +92,67 @@ async fn authenticate(
-        _ => 0,
-    }
-}
-
-fn verify_password(password: &str, secret: &Secret) -> AuthenticationResult {
-    let mut computed_hash = password.to_owned() + &secret.salt;
-    let mut hasher = Sha256::new();
 
-    hasher.input_str(&computed_hash);
-    computed_hash = hasher.result_str();
-
-    if computed_hash.eq(&secret.hash) {
-        AuthenticationResult::Success
-    } else {
-        AuthenticationResult::Failed
-    }
-}
 
 #[derive(Clone)]
 struct WebData {
-    connection_pool: PgPool,
+    connection_pool: database::PgPool,
     token_expiration_seconds: u64,
     token_signing_private_key: Vec<u8>,
 }
 
-const TOKEN_EXPIRATION_SECONDS_ENV_VAR: &str = "TOKEN_EXPIRATION";
-const JWT_SIGNING_PRIVATE_KEY_ENV_VAR: &str = "JWT_ECDSA_SIGNING_KEY";
-
-const DEFAULT_TOKEN_EXPIRATION: &str = "300";
+#[derive(Envconfig)]
+struct Config {
+    #[envconfig(from = "DATABASE_URL")]
+    pub db_url: String,
+    #[envconfig(from = "HTTP_PORT", default = "8080")]
+    pub http_port: u16,
+
+    #[envconfig(from = "TOKEN_EXPIRATION", default = "300")]
+    pub jwt_expiration: u64,
+    #[envconfig(from = "JWT_ECDSA_SIGNING_KEY")]
+    pub jwt_signing_key: Option<String>,
+    #[envconfig(from = "ENABLE_JWT", default = "false")]
+    pub enable_jwt: bool,
+}
 
 #[actix_web::main]
 async fn main() -> std::io::Result<()> {
     env_logger::init();
     dotenv().ok();
 
-    let pool = establish_connection();
-    let jwt_expiration = std::env::var(TOKEN_EXPIRATION_SECONDS_ENV_VAR)
-        .unwrap_or(DEFAULT_TOKEN_EXPIRATION.to_string());
-    let pem_data = std::fs::read(
-        std::env::var(JWT_SIGNING_PRIVATE_KEY_ENV_VAR)
-            .expect("JWT_ECDSA_SIGNING_KEY must be set")).unwrap();
-
-    let data = WebData {
-        connection_pool: pool,
-        token_expiration_seconds: jwt_expiration.parse::<u64>().unwrap(),
-        token_signing_private_key: pem_data,
-    };
-
-    HttpServer::new(move || App::new().service(authenticate).data(data.clone()))
-        .bind("127.0.0.1:8081")?
+    // Initialize config from environment variables
+    let config = Config::init_from_env().unwrap();
+    let data : WebData;
+    let app = App::new();
+
+    let pool = database::establish_connection(config.db_url);
+    if config.enable_jwt {
+        data = WebData{
+            connection_pool: pool,
+            token_expiration_seconds: config.jwt_expiration,
+            token_signing_private_key: std::fs::read(config.jwt_signing_key
+                .expect("JWT_ECDSA_SIGNING_KEY must be set")).unwrap(),
+        };
+        // add the JWT service to the web server.
+        app.service(token_authentication).data(data.clone());
+    } else {
+        data = WebData{
+            connection_pool: pool,
+            token_expiration_seconds: 0,
+            token_signing_private_key: Vec::new(),
+        };
+    }
+
+    //todo fix this :) 
+    HttpServer::new(move || App::new()//app
+        .service(password_authentication).data(data.clone()))
```

Comment:
I tried to create `App` before this call to be able to conditionally enable the JWT service according to the config but it fails to compile at the moment

Reply:
Yes, this isn't going to work from what I know.

You can use the `Condition::new` thing. Or put in optional web data. I combined this for my oauth stuff. So when auth is enabled, the conditonal is true and the optional is set.

Reply:
Condition only works for middleware.. 
I am looking at the configuration struct that should help me with that